### ISSUE TITLE
Can we rewrite numpy operators to pytorch operators?

### ISSUE THREAD

**Issue by zzpmiracle** [pytorch/torchdynamo#1312]
Currently, numpy operations cause graph break, while often exist in training / inference code. But many operators have same semantic in torch operators. Like [YOLOP](https://github.com/hustvl/YOLOP/blob/084d0bb19fa2db4ba40a6e1d6b120f0153d2f335/lib/core/evaluate.py#L230), it would get more than 5x speedup if we change all of the possible numpy to torch, so they can run in gpu.
So can we dynamicly rewrite these numpy code to torch code using torchdynamo, rewrite pattern may like:
`x = tensor.numpy(); b = np.a(x)` to  `x = torch.a(tensor); b = x.numpy()`
or rewrite all support numpy operator `y = numpy.a(x)` to `x = torch.tensor(x); y = torch.a(x); y = y.numpy()` and then erase all the redundant `y = y.numpy(); y = torch.tensor(y)`.

**Comment by Chillee** [comment]
I think we definitely *could*, in theory. I don't know if we want to do it.

I think that if we had full numpy compatibility in PyTorch this would be a lot more reasonable (cc: @ezyang )

**Comment by ezyang** [comment]
Yes, in principle we can do it, but we essentially need PrimTorch but for Numpy API. It is a reasonable thing to do, and you may get good results without too much work, but the long tail is very long!

**Comment by fortianyou** [comment]
Hi, @ezyang @Chillee. Do you have any introduction documents to PrimTorch？What is the goal of PrimTorch?

**Comment by ezyang** [comment]
Check out https://dev-discuss.pytorch.org/t/tracing-with-primitives-update-0/577 and the follow up. also cc @mruberry

**Comment by mruberry** [comment]
You might even be able to `import torch as np` for some programs. PyTorch's `torch` module is already very (but not completely) consistent with the `np` module. It's missing some operations (like those in the `np.random` module) and it doesn't support some features (like arrays of strings). A `torch.np` library could write implementations for many (most?) NumPy operators in terms of PyTorch operations. It would want to focus on operations that PyTorch doesn't have already, or operations where the PyTorch signature is slightly different. Then it might be easier to convert NumPy calls to `torch.np` calls.

**Comment by tanyokwok** [comment]
I think the main idea proposed by @zzpmiracle is to automate rewriting NumPy operations into torch operations. We had to trade-off the full NumPy compatibility in PyTorch. IMO, it can be done with `class NumpyVariable(VairableTracker)`. It seems that PrimTorch is designed for other goals. @Chillee @ezyang
